ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1892,12 +2069,332 @@ void shutdown() throws InterruptedException {
+     *
+     * 1. the process is shutting down, so no need to create new group even if we have more iteration for them to run,
+     *    just let existing groups finish.
+     * 2. shouldTerminateCurrentCycle is true, which means there are replicas added to this thread
+     *    or removed from this thread, don't create new group and just let existing groups finish
+     *    so we can later create new groups with the new set of replicas in next cycle.
+     * 3. any group reaches max iteration, which means existing groups should just finish its current work and stop,
+     *    thus no new groups should be created.
+     * @return {@link boolean} returns true if new groups can be created, false otherwise
+     */
+    private boolean shouldCreateNewGroups() {
+      // check if thread is in shutdown or cycle termination is received or max iterations reached
+      if (!running || shouldTerminateCurrentCycle) {
+        logger.trace(
+            "Thread name: {} not creating new groups as thread is in shutdown mode {} or terminate cycle received {}",
+            threadName, !running, shouldTerminateCurrentCycle);
+        return false;
+      }
+
+      // get max iterations by any remote replica group
+      int maxIterationCount = groupIdIterationCountMap.values().stream().max(Integer::compareTo).orElse(0);
+
+      // check if any replica reached max iteration count
+      if (maxIterationCount >= maxIterationsPerGroup) {
+        logger.trace("Thread name: {} not creating new groups as iteration limit {} reached", threadName,
+            maxIterationsPerGroup);
+        return false;
+      }
+      return true;
+    }
+
+    /**
+     * Generates active and stand by replicas with no progress,
+     * checks and filters replicas which are throttled and are present in any iteration group
+     * @param activeReplicas {@link List}
+     * @param standbyReplicasWithNoProgress {@link List}
+     * @param remoteReplicas {@link List}
+     * @param standByGroupId {@link int}
+     * @param groupId {@link int}
+     */
+    private void fillActiveAndStandByNoProgressReplicas(List<RemoteReplicaInfo> activeReplicas,
+        List<RemoteReplicaInfo> standbyReplicasWithNoProgress, List<RemoteReplicaInfo> remoteReplicas,
+        int standByGroupId, int groupId) {
+      // this will contain all replicas which are available for iteration
+      Set<RemoteReplicaInfo> nonIteratingNonPendingRemoteReplicas = new LinkedHashSet<>(remoteReplicas);
+
+      // remove all throttled replicas
+      remoteReplicas.forEach(remoteReplicaInfo -> {
+        if (replicaToThrottledTill.containsKey(remoteReplicaInfo)
+            && replicaToThrottledTill.get(remoteReplicaInfo) > time.milliseconds()) {
+          nonIteratingNonPendingRemoteReplicas.remove(remoteReplicaInfo);
+          throttleCount.inc();
+        }
+      });
+
+      // remove all replicas of corresponding group id if group is in iteration
```

Comment:
I don't quite understand this logic here. At the beginning, the replicas in a remote node are mapped to group-id. Then here in the fill(...), if that group-id is in-flight, then you rm it from this non-iter-non-xxx list. Wont this list be empty after this ?
is it possible that the some replicas belonging to a group-id are inflight and others not  ? Don't we track in-flight at a group level ?